core/src/main/java/org/apache/calcite/rel/mutable/MutableScan.java

```diff
@@ -36,11 +36,11 @@ public static MutableScan of(TableScan scan) {
   @Override public boolean equals(Object obj) {
     return obj == this
         || obj instanceof MutableScan
-        && rel.equals(((MutableScan) obj).rel);
+        && rel.getTable().equals(((MutableScan) obj).rel.getTable());
   }
```

Comment:
The MaterializedViewSubstitutionVisitor seems should not be used directly.

Reply:
Agree with you on this. However the method `MaterializedViewSubstitutionVisitor.go` is declared public and the correctness of the inner implementation should be ensured, I think. Please reconsider this issue, thanks.

Reply:
Shall we make `MaterializedViewSubstitutionVisitor` non public, if it should not be used directly?

Reply:
We can make it in package scope.

Reply:
If we make `MaterializedViewSubstitutionVisitor` non public, do we still need to update `hashCode` and `equals`, although it's harmless.

Reply:
I prefer updating `hashCode` and `equals` with a reason that we should not leave a bug here. And any other opinion on this?

Reply:
Personally, I would perfer to make `MaterializedViewSubstitutionVisitor` non public and keep `hashCode` and `equals` unchanged.
Or you can compare using the digest and row type of the rel object inside `MutableScan`, just like the HepPlanner does, how do you think about that?

https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/plan/hep/HepPlanner.java#L915

Reply:
Yes, that's an option.